File: 12_StringPathInMatrix/StringPathInMatrix.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <algorithm>
#include <ctime>
using namespace std;

bool hasPathCore(const char *matrix, int rows, int cols, int row, int col, const char *str, int &pathLength, bool *visited)
{
    if (str[pathLength] == '\0')
        return true;
    bool hasPath = false;
    if (row >= 0 && row < rows && col >= 0 && col < cols && matrix[row * cols + col] == str[pathLength] && !visited[row * cols + col])
    {
        pathLength++;
        visited[row * cols + col] = true;
        hasPath = hasPathCore(matrix, rows, cols, row, col - 1, str, pathLength, visited) || hasPathCore(matrix, rows, cols, row - 1, col, str, pathLength, visited) || hasPathCore(matrix, rows, cols, row, col + 1, str, pathLength, visited) || hasPathCore(matrix, rows, cols, row + 1, col, str, pathLength, visited);
        if (!hasPath)
        {
            pathLength--;
            visited[row * cols + col] = false;
        }
    }
    return hasPath;
}

bool hasPath(char *matrix, int rows, int cols, char *str)
{
    if (matrix == nullptr || rows < 1 || cols < 1 || str == nullptr)
        return false;
    bool *visited = new bool[rows * cols];
    fill(visited, visited + rows * cols, false);
    int pathLen = 0;
    for (int row = 0; row < rows; row++)
    {
        for (int col = 0; col < cols; col++)
        {
            if (hasPathCore(matrix, rows, cols, row, col, str, pathLen, visited))
                return true;
        }
    }
    delete[] visited;
    return false;
}

File: 12_StringPathInMatrix/StringPathInMatrix.py
class Solution:
    def hasPath(self, matrix, rows, cols, path):
        if matrix is None or path is None or rows < 1 or cols < 1:
            return False
        visited = [False] * (rows * cols)
        pathLength = 0
        for row in range(rows):
            for col in range(cols):
                if self.hasPathCore(matrix, row, rows, col, cols, pathLength,
                                    path, visited):
                    return True
        return False

    def hasPathCore(self, matrix, row, rows, col, cols, pathLength, path,
                    visited):
        if pathLength == len(path):
            return True
        hasPath = False
        if row >= 0 and row < rows and col >= 0 and col < cols and matrix[
                row * cols +
                col] == path[pathLength] and not visited[row * cols + col]:
            pathLength += 1
            visited[row * cols + col] = True
            hasPath = self.hasPathCore(
                matrix, row + 1, rows, col, cols,
                pathLength, path, visited) or self.hasPathCore(
                    matrix, row - 1, rows, col, cols,
                    pathLength, path, visited) or self.hasPathCore(
                        matrix, row, rows, col + 1, cols, pathLength,
                        path, visited) or self.hasPathCore(
                            matrix, row, rows, col - 1, cols, pathLength, path,
                            visited)
            if not hasPath:
                pathLength -= 1
                visited[row * cols + col] = False
        return hasPath
